decryptor/mysql/response_proxy.go

```diff
@@ -477,64 +514,230 @@ func (handler *Handler) processBinaryDataRow(ctx context.Context, rowData []byte
+			value, err = handler.processFixedSizeNumberField(ctx, i, fields[i], rowData[pos:pos+2])
+			if err != nil {
+				return nil, err
+			}
+			output = append(output, value...)
 			pos += 2
 			continue
 
 		case TypeInt24, TypeLong:
-			output = append(output, rowData[pos:pos+4]...)
+			value, err = handler.processFixedSizeNumberField(ctx, i, fields[i], rowData[pos:pos+4])
+			if err != nil {
+				return nil, err
+			}
+			output = append(output, value...)
 			pos += 4
 			continue
 
 		case TypeLongLong:
-			output = append(output, rowData[pos:pos+8]...)
+			value, err = handler.processFixedSizeNumberField(ctx, i, fields[i], rowData[pos:pos+8])
+			if err != nil {
+				return nil, err
+			}
+			output = append(output, value...)
 			pos += 8
 			continue
 
 		case TypeFloat:
-			output = append(output, rowData[pos:pos+4]...)
+			value, err = handler.processFixedSizeNumberField(ctx, i, fields[i], rowData[pos:pos+4])
+			if err != nil {
+				return nil, err
+			}
+			output = append(output, value...)
 			pos += 4
 			continue
 
 		case TypeDouble:
-			output = append(output, rowData[pos:pos+8]...)
+			value, err = handler.processFixedSizeNumberField(ctx, i, fields[i], rowData[pos:pos+8])
+			if err != nil {
+				return nil, err
+			}
+			output = append(output, value...)
 			pos += 8
 			continue
 
-		case TypeDecimal, TypeNewDecimal,
-			TypeBit, TypeEnum, TypeSet, TypeGeometry:
-			value, _, n, err = LengthEncodedString(rowData[pos:])
-			output = append(output, rowData[pos:pos+n]...)
-			pos += n
+		case TypeDecimal, TypeNewDecimal, TypeBit, TypeEnum, TypeSet, TypeGeometry, TypeDate, TypeNewDate, TypeTimestamp, TypeDatetime, TypeTime:
+			value, n, err = LengthEncodedString(rowData[pos:])
 			if err != nil {
 				handler.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorDecryptorCantDecryptBinary).
 					Errorln("Can't handle length encoded string non binary value")
 				return nil, err
 			}
-			continue
-		case TypeDate, TypeNewDate, TypeTimestamp, TypeDatetime, TypeTime:
-			_, _, n, err = LengthEncodedInt(rowData[pos:])
+			value, err = handler.onColumnDecryption(ctx, i, value)
 			if err != nil {
+				handler.logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorGeneral).
+					WithField("field_index", i).WithError(err).Errorln("Failed to process column data")
 				return nil, err
 			}
-			output = append(output, rowData[pos:pos+n]...)
+			output = append(output, PutLengthEncodedString(value)...)
 			pos += n
 			continue
+
 		default:
 			return nil, fmt.Errorf("found unknown FieldType <type=%d> <name=%s> in MySQL response packet", fields[i].Type, fields[i].Name)
 		}
 	}
 	return output, nil
 }
 
+func (handler *Handler) processFixedSizeNumberField(ctx context.Context, columnIndex int, column *ColumnDescription, encoded []byte) ([]byte, error) {
+	var value []byte
+	var err error
+	// Parse encoded number value into ASCII string (because that's what subscribers expect).
```

Comment:
hm... It good if client's driver sends numbers only using SQL and as strings. Then yes, our proxy work with numbers as with strings. I don't remember good, but probably mysql driver can send values with prepared statement as binary values. And then they will come to us as binary values, not strings. And will be encrypted as binaries. And then if they will back to user as result on some query, probably they will be encoded in text format and proxy will not decrypt them. Or we should encode all formats to one (binary/text to text or binary/text to binary), decrypt/encrypt and then encode to format which we took from client/db...

Reply:
> It good if client's driver sends numbers only using SQL and as strings.

Yeah, that's bothered me as well since binary protocol is a thing, and I'm not that that we handle it at all for MySQL. The integration tests pass, but that's not a guarantee it will work.

The code chunk here does exactly what you say for the decryption path: text is handled as text, binary data is converted into text for processing and then encoded back to the same binary form as in the response sent. However, the encryption path expects and works only with text, as I understand.

Reply:
create please task in phabricator to do it in the future, to separate cases when data encrypted with text format and queried and returned in binary and vice versa with describing the place where it should be done and it's okay for now imho

Reply:
> create please task in phabricator to do it in the future

Tracked in T1757.